File: adress_book_back/adress_book_back/Services/DepartmentService.cs
using adress_book_back.Context;
using adress_book_back.Interfaces;
using adress_book_back.Models;
using adress_book_back.Response;
using Microsoft.EntityFrameworkCore;

namespace adress_book_back.Services
{
    public class DepartmentService:IDepartmentService
    {
        private readonly DBContext _dbContext;

        public DepartmentService(DBContext dbContext)
        {
            _dbContext = dbContext;
        }

        // GET: Paginated Departments
        public async Task<ServiceResponse> GetDepartments(int page = 1, int pageSize = 10)
        {
            var response = new ServiceResponse();
            try
            {
                var query = _dbContext.Departments.AsNoTracking();
                var total = await query.CountAsync();

                var departments = await query
                    .OrderBy(d => d.Id)
                    .Skip((page - 1) * pageSize)
                    .Take(pageSize)
                    .ToListAsync();

                response.State = true;
                response.Data.Add(total);
                response.Data.Add(page);
                response.Data.Add(pageSize);
                response.Data.Add(departments);
            }
            catch (Exception ex)
            {
                response.State = false;
                response.ErrorMessage = $"Error retrieving Departments: {ex.Message}";
            }

            return response;
        }

        // GET: All Departments
        public async Task<ServiceResponse> GetAllDepartments()
        {
            var response = new ServiceResponse();
            try
            {
                var departments = await _dbContext.Departments.AsNoTracking().ToListAsync();
                response.State = true;
                response.Data.Add(departments);
            }
            catch (Exception ex)
            {
                response.State = false;
                response.ErrorMessage = $"Error retrieving Departments: {ex.Message}";
            }

            return response;
        }

        // DELETE: Department by Id
       
        public async Task<bool> Delete(int id)
        {
            var department = await _dbContext.Departments.FirstOrDefaultAsync(x => x.Id == id);
            if (department == null) return false;

            var isUsed = await _dbContext.Addresses.AnyAsync(a => a.DepartmentId == id);
            if (isUsed) return false;

            _dbContext.Departments.Remove(department);
            await _dbContext.SaveChangesAsync();

            return true;
        }

        // SAVE: Insert or Update Department
        public async Task<ServiceResponse> Save(Department department)
        {
            var response = new ServiceResponse();
            try
            {
                if (department.Id == 0)
                    _dbContext.Departments.Add(department);
                else
                    _dbContext.Departments.Update(department);

                await _dbContext.SaveChangesAsync();

                response.State = true;
                response.Data.Add(department);
            }
            catch (Exception ex)
            {
                response.State = false;
                response.ErrorMessage = $"Error saving Department: {ex.Message}";
            }

            return response;
        }
    }
}

File: address_book_dashboard/src/app/core/services/DepartmentService.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { environment } from '../../../environments/environment';
@Injectable({
  providedIn: 'root'
})
export class DepartmentService {
  readonly APIUrl = environment.backApiUrl + 'Departments/';

  constructor(private http: HttpClient) { }

  /** GET: Get paginated Departments */
  getDepartments(page: number = 1, pageSize: number = 20) {
    return this.http.get(`${this.APIUrl}GetDepartments?page=${page}&pageSize=${pageSize}`);
  }

  /** GET: Get All Departments */
  getAllDepartments() {
    return this.http.get(`${this.APIUrl}GetAllDepartments`);
  }

  /** POST: Add or update Department */
  saveDepartment(department: any) {
    return this.http.post(`${this.APIUrl}Save`, department);
  }

  /** DELETE: Delete a Department by Id */
  deleteDepartment(id: number) {
    return this.http.delete(`${this.APIUrl}Delete/${id}`);
  }
}
